src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -280,29 +295,173 @@ def process_input(conf):
+    if _check_file_exists(loader_common_names.SAMPLE_PREPARED_KEY, meta, sample_prepared_file):
+        print(f"Skip generating sample node tree for {upa} as it already exists")
+        return
+
+    try:
+        node_data = _retrieve_node_data(sample_ret['node_tree'])
+    except BadNodeTreeError as e:
+        raise ValueError(f"Error retrieving node data for {upa}") from e
+
+    with open(sample_prepared_file, "w", encoding="utf8") as json_file:
+        json.dump(node_data, json_file, indent=2)
+
+    # write sample file and prepared sample node file name back to the meta file
+    meta[loader_common_names.SAMPLE_FILE_KEY] = sample_file_name
+    meta[loader_common_names.SAMPLE_PREPARED_KEY] = sample_prepared_name
+    with open(metafile, "w", encoding="utf8") as json_file:
+        json.dump(meta, json_file, indent=2)
+
 
-        cfn = os.path.join(conf.job_dir, "workdir/tmp", upa)
-        # upa file is downloaded to cfn
-        conf.asu.get_assembly_as_fasta({"ref": upa.replace("_", "/"), "filename": upa})
+def _check_file_exists(
+        file_key_name: str,
+        metadata: dict[str, Any],
+        file_path: str
+) -> bool:
+    # check if file exists and if the metadata matches the file name
 
-        # each upa in output_dir as a seperate directory
-        dstd = os.path.join(conf.pth, upa)
-        os.makedirs(dstd, exist_ok=True)
+    return (file_key_name in metadata and
+            metadata[file_key_name] == Path(file_path).name and
+            os.path.isfile(file_path))
 
-        dst = os.path.join(dstd, f"{upa}.fa")
-        # Hard link .fa file from job_dir to output_dir in WS
-        os.link(cfn, dst)
 
-        metafile = os.path.join(dstd, f"{upa}.meta")
-        # save meta file with relevant object_info
-        with open(metafile, "w", encoding="utf8") as json_file:
-            json.dump(_process_object_info(obj_info, genome_upa), json_file, indent=2)
+def _retrieve_sample(conf: Conf, upa: str) -> dict[str, Any] | None:
+    # retrieve sample data from sample service
 
-        print("Completed %s" % (upa))
+    # retrieve data links associated with upa
+    links_ret = conf.ss.get_data_links_from_data({"upa": upa.replace("_", "/")})
+
+    data_links = links_ret['links']
+    if not data_links and conf.ignore_no_sample_error:
+        print(f"No sample data links found for {upa}")
+        return
+
+    # there should only be one data link for each upa
+    if len(data_links) != 1:
+        raise ValueError(f"Expected 1 data link for {upa}, got {len(data_links)}")
+
+    # retrieve sample data and save to file
+    sample_id = data_links[0]['id']
+    sample_ret = conf.ss.get_sample_via_data({"upa": upa.replace("_", "/"),
+                                              "id": sample_id,
+                                              "version": data_links[0]["version"]})
+    if not sample_ret:
+        raise ValueError(f"Retrieved empty sample data for {upa}")
```

Comment:
This should be impossible unless something is very very wrong with the sample service

Reply:
👍

Reply:
:+1: